### PR TITLE
Use EXIFTool Available to also Generate ImagePreviews

### PR DESCRIPTION
**Pull request by Urmel**
Hi @nemethviktor ,

I created a first version.

Open topics / Questions.
* Preview logic now is to per default to use a preview (instead of as previously the full image).
  **--> do we want it this way or should normally the full image (when storing in DEs more memory needed) be used?**
* No separate exifToll used yet
* Error message not finalized yet (exception details should go to log, localized error be shown)
* No storing / caching in DEs, yet
* No locking mechanism yet (to protect against concurrent write operation)
  **--> is there one with the previous preview generation code?**
* Only for the main list view, editform and apicallthing are not changed, yet

### CODE REVIEW

**Comment by nemethviktor** [comment]
Hey, awesome work as always! :)

Okay re your Qs:
- I think if people have large monitors, previews looks meh. Mine isn't even a 4K monitor (just 1440p) and previews, espc for jpg files look like they are stretched beyond crazy. For anyone with a 4K or 5K monitor these might look dreadful. 
- Memory shouldn't really be an issue, espc if we don't pre-cache files. ET (at the moment) extracts a full-size preview into JPGs, which aren't particularly huge. Alternatively we can dig a bit to modify the ET config file to force it to resize to something (like max 5K eqv., but even that's too large because our preview window would never be likely that big, so ... half of 5K max or something.)
- As for concurrent writes, hypothetically that can't happen. Not inside the app anyway. I'm a bit lazy to start digging but there is a read-queue-lock logic already in place, which should prevent files being even read while in an ongoing-write-queue. Alternatively the form-lock, which comes up at the load-time of a folder can be extended to also apply to write-time.

**Comment by Urmel** [comment]
Hi @nemethviktor,

first of all - sorry for not responding for such a long time. It's been a bit packed...

## Various topics

> .ExifTool_config file there is some perl code in it related to this. Basically what it does is that it instructs exifTool to look for various types of embedded previews and extract (and rotate) the first one found.

Cool. Thanks for the hint. Do I understand this correctly that the exiftool just reads this cfg file as it sits in the same directory? We do not instruct it to do so (at least I did not find it)...

> That however only gets called for non-jpg files. If Windows can (on its own) decipher the preview (which it's only capable for certain file formats, jpg being one of them) then it does it on its own. Otherwise as a fallback it gets ET to pull the "GTNPreview".

So this "Windows Preview" is generated by reading the image from stream (Image.FromStream), I guess.

> I think if people have large monitors, previews looks meh. Mine isn't even a 4K monitor (just 1440p) and previews, espc for jpg files look like they are stretched beyond crazy. For anyone with a 4K or 5K monitor these might look dreadful.

Yep!!!

> Memory shouldn't really be an issue, espc if we don't pre-cache files. ET (at the moment) extracts a full-size preview into JPGs, which aren't particularly huge. Alternatively we can dig a bit to modify the ET config file to force it to resize to something (like max 5K eqv., but even that's too large because our preview window would never be likely that big, so ... half of 5K max or something.)

How about using 2500 max size (i.e. max of x, y) then (I'll put it in a constant). I guess it would be good to scale all images on load to that size. In memory it doesn't matter if you read a tif or a jpg: the image class itself stores it internally as a bitmap.

I might just give it a shot and click around in a dir with >> 10 images. See what happens to memory... If it gets too bad, the FIFO removal of the cached image might help.

What I also found out is that using the image class based on reading from stream, a lock is kept on the file while the image in memory is not disposed. Question hence would be if we like that. I would rather opt for avoiding the lock or do you see any use of it?


## Locking

> As for concurrent writes, hypothetically that can't happen. Not inside the app anyway. I'm a bit lazy to start digging but there is a read-queue-lock logic already in place, which should prevent files being even read while in an ongoing-write-queue. Alternatively the form-lock, which comes up at the load-time of a folder can be extended to also apply to write-time.

I did the digging part ;) I found one fix that would at least be required (adding locking and unlocking in frmMainApp.lvwFileList_LoadOrUpdate) as I seem to have removed it when getting rid of the EXIF file reading. But shouldn't the same happen in filling the list view in the edit form?


### FilesBeingProcessed

From what I see in DEV branch, FilesBeingProcessed only is filled (via the accessor methods) in ExifRemoveLocationData.
In frmMainApp.lvwFileList_LoadOrUpdate, the list is cleared.
The lock is checked for in:
* frmMainApp.btn_loctToFile_Click
* frmMainApp.tsb_GetAllFromWeb_Click
* frmPasteWhat.btn_OK_Click
* HelperStatic.Exif.ExifRemoveLocationData

Compared to MAIN branch, adding locks should also happen in:
* frmMainApp.lvwFileList_LoadOrUpdate

Compared to MAIN branch, removing should also happen in:
* frmMainApp.lvwFileList_LoadOrUpdate
* HelperStatic.Exif.ExifGetExifFromFilesIncompatibleFileNames (doesn't exist anymore in DEV)
* HelperStatic.Exif.ExifGetExifFromFilesCompatibleFileNames (doesn't exist anymore in DEV)


### FileListBeingUpdated

The bool FileListBeingUpdated is set (normally true & false one each):
* frmMainApp.btn_loctToFile_Click
* frmEditFileData.btn_OK_Click
* frmPasteWhat.btn_OK_Click (+one more false)
* HelperStatic.Exif.ExifRemoveLocationData
* frmMainApp.lvwFileList_LoadOrUpdate (only false)

It is checked for in (together with FilesAreBeingSaved):
* FrmMainApp.FrmMainApp_FormClosing
* FrmMainApp.lvw_FileList_KeyDown
* FrmMainApp.tsb_SaveFiles_Click
* FrmMainApp.tmi_File_SaveAll_Click
* HelperStatic.FSO.FsoCheckOutstandingFiledataOkayToChangeFolderAsync

No difference compared to MAIN branch.


### FilesAreBeingSaved

The bool FilesAreBeingSaved is set
* (true and false) in HelperStatic.ExifWriteExifToFile
* (only false) FrmMainApp.FrmMainApp_FormClosing
* (only false) FrmMainApp.lvw_FileList_KeyDown
* (only false) FrmMainApp.tmi_File_SaveAll_Click
* (only false) FrmMainApp.tsb_SaveFiles_Click

It is checked for in (together with FileListBeingUpdated) and then (see above) set to false:
* FrmMainApp.FrmMainApp_FormClosing
* FrmMainApp.lvw_FileList_KeyDown
* FrmMainApp.tmi_File_SaveAll_Click
* FrmMainApp.tsb_SaveFiles_Click
* HelperStatic.FSO.FsoCheckOutstandingFiledataOkayToChangeFolderAsync

No difference compared to MAIN branch.

**Comment by nemethviktor** [comment]
Hey, no worries - a very quick reply now and I'll write a proper one tomorrow.
I started working on moving away from. The datatables to the DE logic and it's proceeding at a good place so for the time being may I ask you don't work on the code purely because once the DT/DE switch is checked in there'd be a lot of code changes that might require rewriting stuff if you work on it now 
I'll try to incorporate the Image Preview stuff in it somehow:)

**Comment by nemethviktor** [comment]
Okay so I promised to get back to you in a bit better detail. To start with I'd ignore anything that's in the Main branch (in favour of anything that _is_ in the Dev branch, but also see my other comment on the 'canary' file in the other thread).

Now for the Q&As:

### Exiftool

**exiftool cfg file**  - yes it has to sit in the same folder as exiftool itself and it's handled automatically. This is a feature of ET and you'll find if you remove the cfg file the whole shebang will break :) - ET is written in Perl, which I know nothing about at a language but the cfg file is from what I understand a perl file of sorts that allows the users to customize ET's behaviour to very fine details. If you're curious about the details refer to the actual ET website and documentation as it's easier to read there than me trying to explain.

### Images

**"Windows Preview" of images** - yes, Windows creates the previews for the more simple image types. It's the same thing what you see in Windows Explorer if you ever use it. We are using it because it's more efficient than having ET to do extra tasks.

**How about using 2500 max size (i.e. max of x, y)** - either that or a % of monitor width or app width. (monitor is more fixed I guess)

**might give it a shot and click around in a dir with >> 10 images** - I'd say 1000 pics is realistic. Seriously.

### Locking

Locking is only used when writing the file. The `Edit Form` doesn't write to the file, it merely queues up data in whichever "stage" (1/2/3) depending on where in the process the user is. Only when we save the file there is actual locking going on. 
The current logic could be "improved" (read: reconsidered/changed/amended/etc) perhaps by locking the app itself during write (like it is locked during read [entering a folder], however the downside of that would be preventing users moving around folders while a write op is ongoing, as, at the moment one can save the images _and_ move to another folder at the same time.

Regardless of which version you're looking at (as in dev, which uses DTs or canary zip, which uses DEs) the logic is that from Stage3 the data goes into the file itself. Once the write has taken place the file in question is removed from Stage3, therefore we can't really allow anything to write into Stage3 while there's a write-op in process as that could interfere with it.

**Comment by nemethviktor** [comment]
hi @Urmel ,

Btw, had a look at the code in this one (69) against the zip file/canary, I am thinking I'll check in my changes into Dev and this (69) would be quite simple to rebase because the changes, while numerous in the zip don't particularly affect the code in this one? (apart from auto-formatting by resharper but relatively little meaningful change in the particular areas that you edited.
Thoughts?

**Comment by nemethviktor** [comment]
Still hold the horses - some more updates are coming into the dev branch that will be a bit bigger (gotta separate out a lot of classes and files because of the cmd-based request in the other thread, so i can't have everything in one helper.cs file)
I'll give a heads-up about what changed and how to merge. Again since your codes in this PR didn't have very significant changes but more just additions, it shouldn't be too bad :)

File: HelperStatic.Generic.cs [hunk 1]
```diff
@@ -744,17 +738,130 @@ internal static bool GenericLockCheckLockFile(string fileNameWithoutPath)
...
+    /// 
+    /// Either read the read image itself or a preview read via exif tool or created
+    /// via downscaling the read image. Note that using original size uses much more
+    /// memory.
+    /// </summary>
+    /// <param name="fileNameWithPath">Filename w/ path to the image</param>
+    /// <param name="exifTool">The ExifTool instance to use for extraction</param>
+    /// <param name="usePreviewIfAvailable">Whether to use preview (resolution)</param>
+    /// <returns></returns>
+    internal static async Task<(Image, string)> GetImage(string fileNameWithPath,
```
Comment (nemethviktor) [inline HelperStatic.Generic.cs:815]: In the new world this would likely sit in `HelperExifReadGetImagePreviews.cs`

File: FrmMainApp.cs [hunk 2]
```diff
@@ -1605,23 +1605,28 @@ private async Task lvw_HandleSelectionChange()
...
-                                                           .Text);
+                string fileNameWithPath = Path.Combine(
+                    FolderName + lvw_FileList.SelectedItems[index: 0].Text);
 
                 if (File.Exists(path: fileNameWithPath))
                 {
-                    await HelperStatic.GenericCreateImagePreview(
-                        fileNameWithPath: fileNameWithPath, initiator: "FrmMainApp");
+                    // Todo localized error message in return string
+                    Task<(Image, string)> tImg = HelperStatic.GetImage(fileNameWithPath, _ExifTool, true);
```
Comment (nemethviktor) [inline FrmMainApp.cs:1614]: See below but this will likely go into 'HelperExifReadGetImagePreviews.cs'

File: HelperStatic.Generic.cs [hunk 3]
```diff
@@ -668,14 +669,7 @@ internal static bool GenericLockCheckLockFile(string fileNameWithoutPath)
...
 
-        if (initiator == "FrmMainApp" && frmMainAppInstance != null)
-        {
-            frmMainAppInstance.pbx_imagePreview.Image = null;
-            generatedFileName = Path.Combine(path1: FrmMainApp.UserDataFolderPath, path2: frmMainAppInstance.lvw_FileList.SelectedItems[index: 0]
-                                                                                              .Text +
-                                                                                          ".jpg");
-        }
-        else if (initiator == "FrmMainAppAPIDataSelection" && frmMainAppInstance != null)
+        if (initiator == "FrmMainAppAPIDataSelection" && frmMainAppInstance != null)
```
Comment (nemethviktor) [inline HelperStatic.Generic.cs:672]: Again this is now moved to `HelperExifReadGetImagePreviews.cs`

File: Model/ExifToolWrapper.cs [hunk 4]
```diff
@@ -41,27 +41,47 @@
...
-using System.Threading.Tasks;
 using System.Diagnostics;
 using System.IO;
 using GeoTagNinja;
+using static System.Windows.Forms.VisualStyles.VisualStyleElement;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using System.Collections;
 
 namespace ExifToolWrapper
```
Comment (nemethviktor) [inline Model/ExifToolWrapper.cs:53]: I think this file should be okay apart from having renamed a few things - I don't remember splitting this into anything further.

**Review by nemethviktor — COMMENTED** [review]
Made some comments as to where things have moved but you'll find most of the stuff now sits in `HelperExifReadGetImagePreviews` - hoping I'm not causing you tooooo much headache ;)

**Comment by nemethviktor** [comment]
Ok left some comments as above. Will now have a look at making a cmd-based logic for request 71

**Pull request closed without merging**